crates/core/transaction/src/view.rs

```diff
@@ -94,6 +99,173 @@ impl TransactionView {
     pub fn action_views(&self) -> impl Iterator<Item = &ActionView> {
         self.body_view.action_views.iter()
     }
+
+    /// Acts as a higher-order translator that summarizes a TransactionSummary by consolidating
+    /// effects for each unique address.
+    fn accumulate_summary(summary: &TransactionSummary) -> TransactionSummary {
+        let mut transaction_summary = TransactionSummary::default();
+
+        // Get list of unique addresses
+        let addresses = summary
+            .effects
+            .iter()
+            .map(|e| &e.address)
+            .collect::<BTreeSet<_>>();
+
+        for address in &addresses {
+            // Get unique asset IDs for this address
+            let asset_ids = summary
+                .effects
+                .iter()
+                .filter(|effect| effect.address == **address)
+                .flat_map(|effect| effect.balance.balance.keys())
+                .collect::<BTreeSet<_>>();
+
+            // For each unique asset, collect all proto values
+            let mut all_proto_values = Vec::new();
+
+            for asset_id in asset_ids {
+                let proto_values = summary
+                    .effects
+                    .iter()
+                    .filter(|effect| effect.address == **address)
+                    .filter(|effect| effect.balance.balance.contains_key(asset_id))
+                    .flat_map(|effect| ProtoBalance::from(effect.balance.clone()).values)
+                    .collect::<Vec<_>>();
+
+                all_proto_values.extend(proto_values);
+            }
+
+            let combined_proto = ProtoBalance {
+                values: all_proto_values,
+            };
+
+            // Single conversion to domain type
+            let combined_balance: Balance =
+                combined_proto.try_into().expect("Domain type conversion");
+
+            transaction_summary.effects.push(TransactionEffect {
+                address: (*address).clone(),
+                balance: combined_balance,
+            });
+        }
+
+        transaction_summary
+    }
```

Comment:
```suggestion
    fn accumulate_effects(summary: TransactionSummary) -> TransactionSummary {
        use std::collections::BTreeMap;

        let mut keyed_effects: BTreeMap<AddressView, Balance> = BTreeMap::new();
        for effect in summary.effects {
            *keyed_effects.entry(effect.address).or_default() += effect.balance;
        }
        TransactionSummary {
            effects: keyed_effects
                .into_iter()
                .map(|(address, balance)| TransactionEffect { address, balance })
                .collect(),
        }
    }
```

I think this whole method can lean on the existing sum method for Balance?

Reply:
this works well 👍